source/server-discovery-and-monitoring/server-monitoring.rst

```diff
@@ -682,27 +686,29 @@ The event API here is assumed to be like the standard `Python Event
         try:
             if not connection:
                 setUpConnection()
-                return new ServerDescription from handshake response
-
-            if connection.moreToCome:
-                read next isMaster exhaust response
-                return new ServerDescription
-
-            if previousDescription.topologyVersion:
+                rttMonitor.addSample(connection.handshakeDuration)
+                response = connection.handshakeResponse
+            elif connection.moreToCome:
+                response = read next isMaster exhaust response
+            elif previousDescription.topologyVersion:
                 # Initiate streaming isMaster
+                rttMonitor.run()
                 set connection timeout to connectTimeoutMS+heartbeatFrequencyMS
-                call {isMaster: 1, topologyVersion: previousDescription.topologyVersion, maxAwaitTimeMS: heartbeatFrequencyMS}
-                return new ServerDescription
-
-            # The server does not support topologyVersion.
-            set connection timeout to connectTimeoutMS
-            call {isMaster: 1}
-            return new ServerDescription
+                response = call {isMaster: 1, topologyVersion: previousDescription.topologyVersion, maxAwaitTimeMS: heartbeatFrequencyMS}
+            else:
+                # The server does not support topologyVersion.
+                set connection timeout to connectTimeoutMS
+                start = time()
+                response = call {isMaster: 1}
+                rttMonitor.addSample(time()-start)
+
+            return ServerDescription(response, rtt=rttMonitor.average())
```

Comment:
Apologies for barging, but this seemed like an apt PR to bring this up. I interpret this to mean that we should only be using the RTT from the server monitor. Is it permitted for application handshakes to update RTT?

Reply:
I think that's allowed. In fact, if an application handshake can transition a server from Unknown to Known then it might be required to update the RTT as well (to avoid having a 0 value for average RTT). FYI Python/Java do not update the topology from application handshakes.

Reply:
In that case, I may diverge a little from pseudocode to implement RTT monitoring. The server description on the topology may have an existing running average updated by application handshakes, so I cannot overwrite it.

A related question came to mind. Are we explicitly deciding not to update the topology description when RTT receives a response? There is the note to not mark servers unknown on RTT failure:
> Errors encountered when running a "isMaster" command MUST NOT update the topology.

If it's safe to do so, I could just update the RTT of in the topology description as soon as the RTT monitor thread receives an ismaster reply (including the handshake).

Reply:
> Are we explicitly deciding not to update the topology description when the RTT monitor receives a response?

Yes. We decided not to update the topology to avoid churn and to avoid races (incorrectly going from Unknown to known or vice versa). The RttThread can't just publish new ServerDescriptions the same way the Monitor does without adding races. There would need to be a new method that simply updates a server's average rtt.

Could you describe how your idea would work in more detail?

Reply:
Ah gotcha. I can simply update the server's average RTT then. I can take a stab at noting that in the "Measuring RTT" section if that seems worth calling out.

What I have working now is an RTT monitor thread that behaves close to the pseudocode, except shared ServerDescription in the TopologyDescription is responsible for the RTT tracking. When an RTT ismaster or polling ismaster returns, it will add to the shared ServerDescription's RTT tracking. 

Updating just the RTT in a ServerDescription seems natural in libmongoc. In the current behavior, ServerDescription's are not immutable and are responsible for updating the RTT. When a server monitor finishes a polling ismaster, the topology description is fed the BSON reply + RTT or an error. [This is the code path](https://github.com/mongodb/mongo-c-driver/blob/9f43168cfa94313d9e7aeffa75d95f76fb25db50/src/libmongoc/src/mongoc/mongoc-topology.c#L90-L91) where either a server monitor and application handshake updates the topology description after an ismaster attempt. The optional RTT there is the RTT for that single ismaster. Though, FWIW I'm considering making a ServerDescription immutable in libmongoc. It'd be easier to reason about, and is more in line with the SDAM spec's recommendations. But I think the logic for updating just the RTT could be the same.

Reply:
Made a note in https://github.com/mongodb/specifications/pull/810

Reply:
Thanks, I think that sounds reasonable given that C's ServerDescription is mutable.